src/core/include/mp-units/bits/scaling.h

```diff
@@ -0,0 +1,229 @@
+  template<Magnitude M, typename From>
+  static constexpr Rep scale_from(M, const From& value)
+  {
+    return floating_point_scaling_impl<M{}>::template scale<Rep>(value);
+  };
+
+  template<Magnitude M>
+  static constexpr auto scale(M, const Rep& value)
+  {
+    // for standard floating-point types, the result respresentation is always the same
+    return floating_point_scaling_impl<M{}>::template scale<Rep>(value);
+  }
+};
+
+
+template<Magnitude auto M>
+struct fixed_point_scaling_impl {
+  static constexpr Magnitude auto num = _numerator(M);
+  static constexpr Magnitude auto den = _denominator(M);
+  template<std::integral T>
+  static constexpr auto ratio = [] {
+    using U = long double;
+    return detail::fixed_point<T>(_get_value<U>(M));
+  }();
+
+  template<typename To, typename From>
+    requires std::integral<value_type_t<To>> && std::integral<value_type_t<From>>
+  static constexpr To scale(const From& value)
+  {
+    using U = std::common_type_t<value_type_t<From>, value_type_t<To>>;
+    if constexpr (_is_integral(M)) {
+      return static_cast<To>(static_cast<value_type_t<From>>(value) * _get_value<U>(num));
+    } else if constexpr (_is_integral(_pow<-1>(M))) {
+      return static_cast<To>(static_cast<value_type_t<From>>(value) / _get_value<U>(den));
+    } else {
+      return static_cast<To>(ratio<U>.scale(static_cast<value_type_t<From>>(value)));
+    }
+  }
+};
+
+
+template<typename Rep>
+struct fixed_point_scaling_traits {
+  template<Magnitude M, typename From>
+  static constexpr Rep scale_from(M, const From& value)
+  {
+    return fixed_point_scaling_impl<M{}>::template scale<Rep>(value);
+  };
+
+  template<Magnitude M>
+  static constexpr auto scale(M, const Rep& value)
+  {
+    // for standard integer types, the result respresentation is always the same
+    return fixed_point_scaling_impl<M{}>::template scale<Rep>(value);
+  }
+};
+
+
+template<typename T, typename Other = T>
+inline constexpr auto select_scaling_traits = [] {
+  if constexpr (requires {
+                  // we only check if the traits class is complete, not it's members; we do not want to fall-back
+                  // depending on the argument types
+                  { sizeof(mp_units::scaling_traits<T>) } -> std::integral;
+                }) {
+    // traits class is defined; use that
+    return mp_units::scaling_traits<T>{};
+  } else {
+    // undefined traits class; fall-back to default handling
+    if constexpr (mp_units::treat_as_floating_point<T> || mp_units::treat_as_floating_point<Other>) {
+      return floating_point_scaling_traits<T>{};
+    } else if constexpr (std::is_integral_v<T> ||
+                         (std::is_integral_v<value_type_t<T>> && std::convertible_to<value_type_t<T>, T> &&
+                          std::convertible_to<T, value_type_t<T>>)) {
+      return fixed_point_scaling_traits<value_type_t<T>>{};
+    } else {
+      // placeholder to report failure
+      return std::false_type{};
```

Comment:
Not sure I understand correctly: you suggest `select_scaling_traits` should not be defined for those types? Indeed, that is what I wanted, but somehow, I was unable to make it work. Let me try again, maybe it was due to another bug that I have fixed in the meantime...

Reply:
Yes, I think that having `scaling_traits` specialization for `treat_as_floating_point` and `!treat_as_floating_point` would be a better solution here. It would use interfaces that also the user has to use to make things like this work. We do not need any additional dispatch here.

In case of problems with constraints subsumption, we could expose proper concepts to make it work. 

If there is no way to do it as above, then we could implement this in a similar way to CPOs (Niebloids).

Reply:
The latest stint here now indeed works with a `scaling_traits<From,To>`, where `To` is optional. I provided an implementation for `treat_as_floating_point` (assuming that multiplication with a standard floating-point value is correct for all of those) and another one for "integer-likes" (standard integers, plus those which have a nested `value_type` that is a standard integer, and which further interconverts with that). Note that this does not cover all representation types (those which are neither `treat_as_floating_point` nor integer). That is on purpose. Any non-floating-point representation needs precise knowledge about range and precision, which we don't have. So requiring the user to provide a specialisation is the only right path here IMHO.